fvm/src/engine.rs

```diff
@@ -557,8 +557,97 @@ impl Engine {
             .expect("failed to create available_gas global");
         store.data_mut().avail_gas_global = gg;
 
-        store.limiter(|id| id.kernel.limiter_mut());
+        fn as_wasmtime_limiter<K: Kernel>(
+            data: &mut InvocationData<K>,
+        ) -> &mut dyn wasmtime::ResourceLimiter {
+            unsafe {
+                &mut *(data.kernel.limiter_mut() as *mut _ as *mut WasmtimeLimiter<K::Limiter>)
+            }
```

Comment:
Ugh, I don't know even know what I'm looking at with all this `as *mut _ as *mut` voodoo. 
I reckon if we don't _have_ to use `unsafe` we shouldn't. Isn't there a way to avoid this?

Reply:
I'll add a comment. The outer struct is defined to be "transparent" so this cast is guaranteed to be safe.

Unfortunately, the store expects a _borrowed_ limiter, so I can't just wrap the internal limiter unless I store the wrapper somewhere. I'm cheating here by just casting because I know the "wrapper" is transparent.

Reply:
My worry is that these kind of clever tricks will be normalised across the codebase for getting around finding better ways to design the code. It worked before without using unsafe. We didn't like having to have a limiter on kernel in the first place, but it was a necessary evil to make it available where it had to be. Now it feels like adding insult to injury by having code that at least one person on the team (me) wouldn't know how to write or maintain.  

Normally I'd expect unsafe to be used in places like data structures and the Wasm SDK, where they are wrapped so the users of those components don't have to worry about them; they are written by an expert and air gapped. This feels like a more ad-hoc use that is more of a code smell.